## 1. The NIST Special Publication 800-63B
<p>If you – 50 years ago – needed to come up with a secret password you were probably part of a secret espionage organization or (more likely) you were pretending to be a spy when playing as a kid. Today, many of us are forced to come up with new passwords <em>all the time</em> when signing into sites and apps. As a password <em>inventeur</em> it is your responsibility to come up with good, hard-to-crack passwords. But it is also in the interest of sites and apps to make sure that you use good passwords. The problem is that it's really hard to define what makes a good password. However, <em>the National Institute of Standards and Technology</em> (NIST) knows what the second best thing is: To make sure you're at least not using a <em>bad</em> password. </p>
<p>In this notebook, we will go through the rules in <a href="https://pages.nist.gov/800-63-3/sp800-63b.html">NIST Special Publication 800-63B</a> which details what checks a <em>verifier</em> (what the NIST calls a second party responsible for storing and verifying passwords) should perform to make sure users don't pick bad passwords. We will go through the passwords of users from a fictional company and use Python to flag the users with bad passwords. But us being able to do this already means the fictional company is breaking one of the rules of 800-63B:</p>
<blockquote>
  <p>Verifiers SHALL store memorized secrets in a form that is resistant to offline attacks. Memorized secrets SHALL be salted and hashed using a suitable one-way key derivation function.</p>
</blockquote>
<p>That is, never save users' passwords in plaintext, always encrypt the passwords! Keeping this in mind for the next time we're building a password management system, let's load in the data.</p>
<p><em>Warning: The list of passwords and the fictional user database both contain <strong>real</strong> passwords leaked from <strong>real</strong> websites. These passwords have not been filtered in any way and include words that are explicit, derogatory and offensive.</em></p>

### **Task 1: Instructions**

Load in and inspect the usernames and passwords of the fictional users.

* Load the `pandas` module.

* Load the user data from the file contained in the path `datasets/users.csv` and store it as a DataFrame called `users`.

* Print the number of rows (i.e. users).

* Show the first 12 rows in `users`.

To complete this project, you need to know how to manipulate strings in `pandas` DataFrames and be familiar with regular expressions. Before starting this project we recommend that you have completed the following courses:

* **[Data Cleaning in Python](https://www.datacamp.com/courses/data-cleaning-in-python)**

* **[Regular Expressions in Python](https://www.datacamp.com/courses/data-cleaning-in-python)**

An excellent companion while doing this project is the ***[Working with text data](https://pandas.pydata.org/pandas-docs/stable/text.html)***  At the bottom of that page you'll also find a list of all the string functions `pandas` supports.

In [34]:
# Importing the pandas module
import pandas as pd

# Loading in datasets/users.csv 
users = pd.read_csv('datasets/users.csv')

# Printing out how many users we've got
print(users.shape[0])

# Taking a look at the 12 first users
users.head(12)

982


id         user_name          password
0    1    vance.jennings          joobheco
1    2    consuelo.eaton        0869347314
2    3   mitchel.perkins        fabypotter
3    4    odessa.vaughan         aharney88
4    5    araceli.wilder        acecdn3000
5    6  shawn.harrington           5278049
6    7        evelyn.gay            master
7    8       noreen.hale            murphy
8    9       gladys.ward           lwsves2
9   10   brant.zimmerman  1190KAREN5572497
10  11     leanna.abbott          aivlys24
11  12   milford.hubbard           hubbard

## 2. Passwords should not be too short
<p>If we take a look at the first 12 users above we already see some bad passwords. But let's not get ahead of ourselves and start flagging passwords <em>manually</em>. What is the first thing we should check according to the NIST Special Publication 800-63B?</p>
<blockquote>
  <p>Verifiers SHALL require subscriber-chosen memorized secrets to be at least 8 characters in length.</p>
</blockquote>
<p>Ok, so the passwords of our users shouldn't be too short. Let's start by checking that!</p>

### **Task 2: Instructions**

Flag the passwords that are too short.

* Add the column `length` to `users` which should list the number of characters in each `password`.

* Flag the users with too short passwords by adding the column `users['too_short']` which should be `True` when `users['length']` is less than 8.

* Print the count of the number of users with passwords that are too short.

* Show the first 12 rows in `users`.

To solve this task, you need to be able to figure out the number of characters in each password. Check the `pandas` **[string documentation](https://pandas.pydata.org/pandas-docs/stable/text.html)** for a method that does just that.

In [36]:
# Calculating the lengths of users' passwords
users['length'] = users['password'].str.len()

# Flagging the users with too short passwords
users['too_short'] = users['length'].apply(lambda x: True if x < 8 else False)

# Counting and printing the number of users with too short passwords
print(users['too_short'].sum())

# Taking a look at the 12 first rows
users.head(12)

376


id         user_name          password  length  too_short
0    1    vance.jennings          joobheco       8      False
1    2    consuelo.eaton        0869347314      10      False
2    3   mitchel.perkins        fabypotter      10      False
3    4    odessa.vaughan         aharney88       9      False
4    5    araceli.wilder        acecdn3000      10      False
5    6  shawn.harrington           5278049       7       True
6    7        evelyn.gay            master       6       True
7    8       noreen.hale            murphy       6       True
8    9       gladys.ward           lwsves2       7       True
9   10   brant.zimmerman  1190KAREN5572497      16      False
10  11     leanna.abbott          aivlys24       8      False
11  12   milford.hubbard           hubbard       7       True

## 3.  Common passwords people use
<p>Already this simple rule flagged a couple of offenders among the first 12 users. Next up in Special Publication 800-63B is the rule that</p>
<blockquote>
  <p>Verifiers SHALL compare the prospective secrets against a list that contains values known to be commonly-used, expected, or compromised.</p>
  <ul>
  <li>Passwords obtained from previous breach corpuses.</li>
  <li>Dictionary words.</li>
  <li>Repetitive or sequential characters (e.g. ‘aaaaaa’, ‘1234abcd’).</li>
  <li>Context-specific words, such as the name of the service, the username, and derivatives thereof.</li>
  </ul>
</blockquote>
<p>We're going to check these in order and start with <em>Passwords obtained from previous breach corpuses</em>, that is, websites where hackers have leaked all the users' passwords. As many websites don't follow the NIST guidelines and encrypt passwords there now exist large lists of the most popular passwords. Let's start by loading in the 10,000 most common passwords which I've taken from <a href="https://github.com/danielmiessler/SecLists/tree/master/Passwords">here</a>.</p>

### **Task 3: Instructions**

Load in the list with the 10,000 most common passwords.

* Read in `datasets/10_million_password_list_top_10000.txtas` a Series and put it in the variable `common_passwords`.

* Take a look at the top 20 common passwords.

*Note: The passwords are stored in a plain text file with one password per row. To read this in as a Series you can use the `read_csv` function from pandas but to make it return a Series (rather than a DataFrame) you have to set the arguments `header=None` and `squeeze=True`.*

In [38]:
# Reading in the top 10000 passwords
common_passwords = pd.read_csv('datasets/10_million_password_list_top_10000.txt', squeeze=True, header=None)

# Taking a look at the top 20
common_passwords.head(20)

0        123456
1      password
2      12345678
3        qwerty
4     123456789
5         12345
6          1234
7        111111
8       1234567
9        dragon
10       123123
11     baseball
12       abc123
13     football
14       monkey
15      letmein
16       696969
17       shadow
18       master
19       666666
Name: 0, dtype: object

## 4.  Passwords should not be common passwords
<p>The list of passwords was ordered, with the most common passwords first, and so we shouldn't be surprised to see passwords like <code>123456</code> and <code>qwerty</code> above. As hackers also have access to this list of common passwords, it's important that none of our users use these passwords!</p>
<p>Let's flag all the passwords in our user database that are among the top 10,000 used passwords.</p>

### **Task 4: Instructions**

Flag the `users` passwords that are among the top 10,000 used passwords.

* Flag common `users` passwords by adding the column `users['common_password']` which should be `True` when a password is one of the `common_passwords`.

* Count and print the number of users using common passwords.

* Show the first 12 rows in users.

In [40]:
# Flagging the users with passwords that are common passwords
users['common_password'] = users['password'].isin(common_passwords)

# Counting and printing the number of users using common passwords
print(users[users['common_password']].shape[0])

# Taking a look at the 12 first rows
users.head(12)

129


id         user_name          password  length  too_short  common_password
0    1    vance.jennings          joobheco       8      False            False
1    2    consuelo.eaton        0869347314      10      False            False
2    3   mitchel.perkins        fabypotter      10      False            False
3    4    odessa.vaughan         aharney88       9      False            False
4    5    araceli.wilder        acecdn3000      10      False            False
5    6  shawn.harrington           5278049       7       True            False
6    7        evelyn.gay            master       6       True             True
7    8       noreen.hale            murphy       6       True             True
8    9       gladys.ward           lwsves2       7       True            False
9   10   brant.zimmerman  1190KAREN5572497      16      False            False
10  11     leanna.abbott          aivlys24       8      False            False
11  12   milford.hubbard           hubbard       7       True            False

## 5. Passwords should not be common words
<p>Ay ay ay! It turns out many of our users use common passwords, and of the first 12 users there are already two. However, as most common passwords also tend to be short, they were already flagged as being too short. What is the next thing we should check?</p>
<blockquote>
  <p>Verifiers SHALL compare the prospective secrets against a list that contains […] dictionary words.</p>
</blockquote>
<p>This follows the same logic as before: It is easy for hackers to check users' passwords against common English words and therefore common English words make bad passwords. Let's check our users' passwords against the top 10,000 English words from <a href="https://github.com/first20hours/google-10000-english">Google's Trillion Word Corpus</a>.</p>

### **Task 5: Instructions**

Flag the passwords that are among the 10,000 most common English words.

* Read in `datasets/google-10000-english.txt` as a Series and put it in the variable `words`.

* Flag `users` passwords that are common words by adding the column `users['common_word']` which should be `True` when a `password` is one of the `words`. The comparison should be case-insensitive.

* Count and print the number of users using common words as passwords.

* Show the first 12 rows in `users`.

In [42]:
# Reading in a list of the 10000 most common words
words = pd.read_csv('datasets/google-10000-english.txt', squeeze=True, header=None)

# Flagging the users with passwords that are common words
users['common_word'] = users['password'].str.lower().isin(words)

# Counting and printing the number of users using common words as passwords
print(users[users['common_word']].shape[0])

# Taking a look at the 12 first rows
users.head(12)

137


id         user_name          password  length  too_short  \
0    1    vance.jennings          joobheco       8      False   
1    2    consuelo.eaton        0869347314      10      False   
2    3   mitchel.perkins        fabypotter      10      False   
3    4    odessa.vaughan         aharney88       9      False   
4    5    araceli.wilder        acecdn3000      10      False   
5    6  shawn.harrington           5278049       7       True   
6    7        evelyn.gay            master       6       True   
7    8       noreen.hale            murphy       6       True   
8    9       gladys.ward           lwsves2       7       True   
9   10   brant.zimmerman  1190KAREN5572497      16      False   
10  11     leanna.abbott          aivlys24       8      False   
11  12   milford.hubbard           hubbard       7       True   

    common_password  common_word  
0             False        False  
1             False        False  
2             False        False  
3             False        False  
4             False        False  
5             False        False  
6              True         True  
7              True         True  
8             False        False  
9             False        False  
10            False        False  
11            False        False

## 6. Passwords should not be your name
<p>It turns out many of our passwords were common English words too! Next up on the NIST list:</p>
<blockquote>
  <p>Verifiers SHALL compare the prospective secrets against a list that contains […] context-specific words, such as the name of the service, the username, and derivatives thereof.</p>
</blockquote>
<p>Ok, so there are many things we could check here. One thing to notice is that our users' usernames consist of their first names and last names separated by a dot. For now, let's just flag passwords that are the same as either a user's first or last name.</p>

### **Task 6: Instructions**

Flag passwords that are the same as the users first or last name.

* Extract users first names from `users['user_name']` into the new column `users['first_name']`.

* Similarly, extract last names into the new column `users['last_name']`.

* Add the column `users['uses_name']` which should be `True` when a password is the same as each users' first or last name.

* Count and print the number of users using names as passwords.

* Show the first 12 rows in `users`.

To extract the first and last names you can use the **`.str.extract()`** method. Check out the pandas documentation under **[Extracting substrings](https://pandas.pydata.org/pandas-docs/stable/user_guide/text.html#extracting-substrings)** for more info.

In [44]:
# Extracting first and last names into their own columns
users['first_name'] = users['user_name'].str.extract(r"(^[A-Za-z]+)", expand=False)
users['last_name'] = users['user_name'].str.extract(r"([A-Za-z]+?)", expand=False)

# Flagging the users with passwords that matches their names
users['uses_name'] = (users['first_name'] == users['password']) | (users['last_name'] == users['password'])

# Counting and printing the number of users using names as passwords
print(users['uses_name'].sum())

# Taking a look at the 12 first rows
users.head(12)

22


id         user_name          password  length  too_short  \
0    1    vance.jennings          joobheco       8      False   
1    2    consuelo.eaton        0869347314      10      False   
2    3   mitchel.perkins        fabypotter      10      False   
3    4    odessa.vaughan         aharney88       9      False   
4    5    araceli.wilder        acecdn3000      10      False   
5    6  shawn.harrington           5278049       7       True   
6    7        evelyn.gay            master       6       True   
7    8       noreen.hale            murphy       6       True   
8    9       gladys.ward           lwsves2       7       True   
9   10   brant.zimmerman  1190KAREN5572497      16      False   
10  11     leanna.abbott          aivlys24       8      False   
11  12   milford.hubbard           hubbard       7       True   

    common_password  common_word first_name last_name  uses_name  
0             False        False      vance         v      False  
1             False        False   consuelo         c      False  
2             False        False    mitchel         m      False  
3             False        False     odessa         o      False  
4             False        False    araceli         a      False  
5             False        False      shawn         s      False  
6              True         True     evelyn         e      False  
7              True         True     noreen         n      False  
8             False        False     gladys         g      False  
9             False        False      brant         b      False  
10            False        False     leanna         l      False  
11            False        False    milford         m      False

In [45]:
users[users['uses_name']].head()

id        user_name  password  length  too_short  common_password  \
84    85  dorian.caldwell    dorian       6       True             True   
112  113    juliette.rios  juliette       8      False             True   
122  123     caleb.powers     caleb       5       True            False   
147  148      lara.bright      lara       4       True             True   
151  152  roland.whitaker    roland       6       True             True   

     common_word first_name last_name  uses_name  
84         False     dorian         d       True  
112        False   juliette         j       True  
122        False      caleb         c       True  
147        False       lara         l       True  
151         True     roland         r       True

## 7. Passwords should not be repetitive
<p>Milford Hubbard (user number 12 above), what where you thinking!? Ok, so the last thing we are going to check is a bit tricky:</p>
<blockquote>
  <p>Verifiers SHALL compare the prospective secrets [so that they don't contain] repetitive or sequential characters (e.g. ‘aaaaaa’, ‘1234abcd’).</p>
</blockquote>
<p>This is tricky to check because what is <em>repetitive</em> is hard to define. Is <code>11111</code> repetitive? Yes! Is <code>12345</code> repetitive? Well, kind of. Is <code>13579</code> repetitive? Maybe not..? To check for <em>repetitiveness</em> can be arbitrarily complex, but here we're only going to do something simple. We're going to flag all passwords that contain 4 or more repeated characters.</p>

### **Task 7: Instructions**

Flag passwords that contain 4 or more repeated characters.

* Add the column `users['too_many_repeats']` which should be `True` when a `password` has 4 or more repeated characters.

* Take a look at only the `users` with too many repeats.

This task can be solved using the **`.str.contains()`** method that is described in the pandas **[text documentation](https://pandas.pydata.org/pandas-docs/stable/text.html#testing-for-strings-that-match-or-contain-a-pattern)**. The regexp you need to craft here is a bit tricky, so you may need to revisit this **[video lesson in Regular Expressions in Python.

In [47]:
### Flagging the users with passwords with >= 4 repeats
users['too_many_repeats'] = users['password'].str.contains(r"(.)\1{3}", regex=True)

# Taking a look at the users with too many repeats
users[users['too_many_repeats']]

id         user_name  password  length  too_short  common_password  \
146  147       patti.dixon    555555       6       True             True   
572  573  cornelia.bradley    555555       6       True             True   
644  645       essie.lopez     11111       5       True             True   
798  799       charley.key    888888       6       True             True   
807  808   thurman.osborne  rinnnng0       8      False            False   
941  942    mitch.ferguson    aaaaaa       6       True             True   

     common_word first_name last_name  uses_name  too_many_repeats  
146        False      patti         p      False              True  
572        False   cornelia         c      False              True  
644        False      essie         e      False              True  
798        False    charley         c      False              True  
807        False    thurman         t      False              True  
941        False      mitch         m      False              True

## 8. All together now!
<p>Now we have implemented all the basic tests for bad passwords suggested by NIST Special Publication 800-63B! What's left is just to flag all bad passwords and maybe to send these users an e-mail that strongly suggests they change their password.</p>

### **Task 8: Instructions**

Flag *all* the bad passwords.

* Add the column `users['bad_password']` which should be `True` when a password is bad according to `too_short`, `common_password`, `common_word`, `uses_name`, or `too_many_repeats`.

* Count and print the number of users with bad passwords.

* Show the first 25 bad passwords in `users`.

In [49]:
# Flagging all passwords that are bad
users['bad_password'] = users['too_short'] | users['common_password'] | users['common_word'] | users['uses_name'] | users['too_many_repeats']

# Counting and printing the number of bad passwords
print(users[users['bad_password']].shape[0])

# Looking at the first 25 bad passwords
users[users['bad_password']].head(25)


421


id         user_name   password  length  too_short  common_password  \
5    6  shawn.harrington    5278049       7       True            False   
6    7        evelyn.gay     master       6       True             True   
7    8       noreen.hale     murphy       6       True             True   
8    9       gladys.ward    lwsves2       7       True            False   
11  12   milford.hubbard    hubbard       7       True            False   
13  14     jamie.cochran     310356       6       True            False   
15  16        lorrie.gay    oZ4k0QE       7       True            False   
16  17      domingo.dyer    chelsea       7       True             True   
17  18    martin.pacheco    zvc1939       7       True            False   
18  19     shelby.massey     nickgd       6       True            False   
21  22   leticia.sanford   cocacola       8      False             True   
22  23     jenny.woodard    woodard       7       True            False   
25  26      dianna.munoz      AJ9Da       5       True            False   
26  27      julia.savage     ewokzs       6       True            False   
28  29   joaquin.walters    YyGjz8E       7       True            False   
30  31      rosanna.reid       reid       4       True            False   
34  35      roger.golden    jOYZBs8       7       True            False   
38  39       gus.padilla    wwewwf1       7       True            False   
43  44    vito.nicholson     225377       6       True            False   
45  46      morris.price     NdZ7E6       6       True            False   
47  48      jarred.white      CQB3Z       5       True            False   
48  49      junior.burch      diffo       5       True            False   
51  52      ahmad.hopper  123456789       9      False             True   
52  53   rosario.merrill    y8uM7D6       7       True            False   
56  57      dwayne.hurst    mikeloo       7       True            False   

    common_word first_name last_name  uses_name  too_many_repeats  \
5         False      shawn         s      False             False   
6          True     evelyn         e      False             False   
7          True     noreen         n      False             False   
8         False     gladys         g      False             False   
11        False    milford         m      False             False   
13        False      jamie         j      False             False   
15        False     lorrie         l      False             False   
16         True    domingo         d      False             False   
17        False     martin         m      False             False   
18        False     shelby         s      False             False   
21        False    leticia         l      False             False   
22        False      jenny         j      False             False   
25        False     dianna         d      False             False   
26        False      julia         j      False             False   
28        False    joaquin         j      False             False   
30         True    rosanna         r      False             False   
34        False      roger         r      False             False   
38        False        gus         g      False             False   
43        False       vito         v      False             False   
45        False     morris         m      False             False   
47        False     jarred         j      False             False   
48        False     junior         j      False             False   
51        False      ahmad         a      False             False   
52        False    rosario         r      False             False   
56        False     dwayne         d      False             False   

    bad_password  
5           True  
6           True  
7           True  
8           True  
11          True  
13          True  
15          True  
16          True  
17          True  
18          True  
21          True  
22          True  
25         

## 9. Otherwise, the password should be up to the user
<p>In this notebook, we've implemented the password checks recommended by the NIST Special Publication 800-63B. It's certainly possible to better implement these checks, for example, by using a longer list of common passwords. Also note that the NIST checks in no way guarantee that a chosen password is good, just that it's not obviously bad.</p>
<p>Apart from the checks we've implemented above the NIST is also clear with what password rules should <em>not</em> be imposed:</p>
<blockquote>
  <p>Verifiers SHOULD NOT impose other composition rules (e.g., requiring mixtures of different character types or prohibiting consecutively repeated characters) for memorized secrets. Verifiers SHOULD NOT require memorized secrets to be changed arbitrarily (e.g., periodically).</p>
</blockquote>
<p>So the next time a website or app tells you to "include both a number, symbol and an upper and lower case character in your password" you should send them a copy of <a href="https://pages.nist.gov/800-63-3/sp800-63b.html">NIST Special Publication 800-63B</a>.</p>

###  **Task 9: Instructions**

* Assign a new password to `new_password` that passes the NIST rules you've implemented in this project.

In [51]:
# Enter a password that passes the NIST requirements
# PLEASE DO NOT USE AN EXISTING PASSWORD HERE
new_password = "$9gIgk2mx-+"